Advanced Lift-Drag Plugin (#901)
PX4/sitl_gazebo: src/liftdrag_plugin/Advanced Lift Drag Readme.md

> <h1>Introduction</h1><br>This document describes the new Gazebo plugin for aerodynamic modeling, which updates the physics formulation, adding in quadratic drag, sideforce, moments about all three axes, and body rate derivatives.

Reviewer: ```suggestion
<h1>Introduction</h1><br>This document describes the the `advanced_liftdrag_plugin`. Primary modeling differences from the `liftdrag_plugin` include:
- quadratic formulation for drag
- side force
- flat plate post stall model
- aerodynamic moments about all three axes
- body rate stability derivatives
- actuator control derivatives
```

<h1>Notation</h1>
All the variables representing aerodynamic coefficients and derivatives are named according to the following convention: <br>
<ul>
    <li>The first letter of these variable names is a capital C.</li>
    <li>The next letter(s) represent the force or moment that this variable relates to.</li>
    <ul>
        <li>If this variable pertains to a force, the next letter is a capital L, D, or Y, representing <b>L</b>ift, <b>D</b>rag, or sideforce (force in the <b>Y</b>- direction) respectively.</li>
        <li>If this variable pertains to a moment, the next letters are “ell”, “em”, or “en”, all lowercase, representing roll moment, pitch moment, and yaw moment respectively. This was chosen to avoid confusion between CL, the coefficient of lift, and Cell, the coefficient of roll moment.</li>
    </ul>
    <li>The next letter(s) represent what the coefficient is being differentiated with respect to.</li>
    <ul>

>         <li>0: represents zero-angle-of-attack or zero-lift condition (not a differential)</li>

Reviewer: ```suggestion
        <li>0: represents the respective coefficient with zero angle of attack, body rates, and actuator deflection (not a differential)</li>
```

        <li>a: angle of attack, alpha</li>
        <li>b: sideslip angle, beta</li>
        <li>ctrl: control surface deflection</li>
        <li>p: non-dimensionalized roll rate</li>
        <li>q: non-dimensionalized pitch rate</li>

>         <li>r: non-dimensionalized yaw rate</li>

Reviewer: could you please either
1. add definitions for the non-dimensionalizations (as they differ between lateral-directional and longitudinal), e.g. $p = pb/(2V)$, where $p$ is the body roll rate, $b$ is wing span and $V$ is total (true) airspeed.
2. add the definitions and additionally put e.g. a "hat" symbol over the non-dimensionalized rates to disambiguate $\hat{p}$

    </ul>
 </ul>
One potentially confusing item in the code is the abbreviation “CDp”. CDp could be used to refer to a body rate derivative (the change in coefficient of drag with respect to the non-dimensionalized roll rate) or parasitic drag. This code uses it to refer to the body rate derivative: the parasitic drag (also known as “zero-lift drag”) is referred to as CD0 in the code. <br>
<h1>Coordinate Frames</h1>
Gazebo uses an ENU (east-north-up) reference frame, while the AVL equations of motion assume an NED (north-east-down) frame. Both of these frames are mapped onto the aircraft assuming that it is traveling along the X-axis of the frame (east in ENU, north in NED). What this means in practice is that the ENU frame has a Y axis out the left wing and a Z axis pointing upward, while the NED frame has a Y axis pointing out the right wing and a Z axis pointing downward. As such, the angles, forces, and moments need to be translated between the frames.

> As of August 2022, the advanced lift-drag plugin computes all forces in the stability frame, and all moments in the body frame. The stability frame is a reference frame oriented such that the X-axis is pointed opposite to the projection of the relative wind vector onto the plane of symmetry of the aircraft. In other words, compared to the body frame, the stability frame has been rotated about the Y-axis by an angle ⍺, the angle of attack. This is important, since the lift and drag are computed in this frame. The following Stack Overflow post (https://aviation.stackexchange.com/questions/70577/under-what-circumstances-is-the-aircraft-stability-axis-system-used-in-either) contains a diagram showing the stability, body, and wind axes. <br>

Reviewer: ```suggestion
The advanced lift-drag plugin computes all forces in the stability frame, and all moments in the body frame. The stability frame is a reference frame oriented such that the X-axis is pointed opposite to the projection of the relative wind vector onto the plane of symmetry of the aircraft. In other words, compared to the body frame, the stability frame has been rotated about the Y-axis by an angle ⍺, the angle of attack. This is important, since the lift and drag are computed in this frame. The following Stack Overflow post (https://aviation.stackexchange.com/questions/70577/under-what-circumstances-is-the-aircraft-stability-axis-system-used-in-either) contains a diagram showing the stability, body, and wind axes. <br>
```

> As of August 2022, the advanced lift-drag plugin computes all forces in the stability frame, and all moments in the body frame. The stability frame is a reference frame oriented such that the X-axis is pointed opposite to the projection of the relative wind vector onto the plane of symmetry of the aircraft. In other words, compared to the body frame, the stability frame has been rotated about the Y-axis by an angle ⍺, the angle of attack. This is important, since the lift and drag are computed in this frame. The following Stack Overflow post (https://aviation.stackexchange.com/questions/70577/under-what-circumstances-is-the-aircraft-stability-axis-system-used-in-either) contains a diagram showing the stability, body, and wind axes. <br>

Reviewer: > The following Stack Overflow post (https://aviation.stackexchange.com/questions/70577/under-what-circumstances-is-the-aircraft-stability-axis-system-used-in-either) contains a diagram showing the stability, body, and wind axes.

Could we find a less temporary reference for this than a stack overflow post? -- e.g. there should be a book that displays this figure (e.g. Stengel, Flight Dynamics; Anderson, Fundamentals of Aerodynamics) -- would even be nice to put the full image in this readme, and cite the source.

<h1>Getting the Aerodynamic Parameters</h1>
<h2>AVL</h2>

> AVL, also known as Athena Vortex Lattice, was used to obtain the aerodynamic parameters and derivatives needed for the advanced lift-drag plugin. AVL uses a vortex-lattice method to compute the forces and moments acting on an aircraft at any arbitrary flight condition, and can also compute the stability and control derivatives needed to create a linearized simulation of an aircraft about a trim condition. <br> To use AVL, you first need to create a file containing your aircraft configuration. The AVL User Primer (https://web.mit.edu/drela/Public/web/avl/AVL_User_Primer.pdf) explains how to do this. Another useful resource is the variety of example files that comes with AVL: modifying one of those to suit your needs is typically much easier than writing an AVL file from scratch. <br> Once you have created your aircraft input file, open up AVL, and load your aircraft using the LOAD command. Go to the operations menu with the OPER command; from here, typing in X and hitting the Enter/Return key should execute a flow computation. Once you are satisfied with the computation, use the ST command to get derivatives in the stability axes and the SB commands to get derivatives in the body axes. Note that you will need derivatives from both of these commands to fill out the advanced lift-drag plugin. <br> One important fact to note. This plugin treats each individual control surface as its own surface; it cannot combine them the way AVL can. As such, if you are trying to determine the control derivatives for your aircraft’s ailerons, you need to determine the control derivatives for the left and the right ailerons individually- doing them both at once may not work. <br> <br>Make sure to sanity-check your results! While AVL is generally reliable, there are some cases where it has produced erroneous results. For example, when I ran an R/C aircraft after something large (e.g. the Boeing 737 example model), AVL used the wrong reference point; as a result, several derivatives came out enormously larger than they should have.

Reviewer: > For example, when I ran an R/C aircraft after something large (e.g. the Boeing 737 example model), AVL used the wrong reference point; as a result, several derivatives came out enormously larger than they should have.

For example, running an R/C aircraft after something large (e.g. the Boeing 737 example model) without resetting the reference point can result in very large, incorrect derivatives.

^maybe further describing what this reference point is in the example would also help -- even if just a short couple of words for context.

$$C_L  =  C_{L0} + C_{Lα} α + C_{Lp} p+ C_{Lq} q+ C_{Lr} r+ \sum_{x=1}^{CS} C_{L,ctrl,x} \delta_{ctrl,x}$$
$$C_D  =  C_{D0} + \frac{{{C_L}} ^ {2}}{π ARe} + C_{Dp} p+ C_{Dq} q+ C_{Dr} r+ \sum_{x=1}^{CS} C_{D,ctrl,x} \delta_{ctrl,x}$$
$$C_Y  =  C_{Y β} β + C_{Yp} p+ C_{Yq} q+ C_{Yr} r+ \sum_{x=1}^{CS} C_{Y,ctrl,x} \delta_{ctrl,x}$$
$$C_ℓ  =  C_{ℓ β} β + C_{ℓp} p+ C_{ℓq} q+ C_{ℓr} r+ \sum_{x=1}^{CS} C_{ℓ,ctrl,x} \delta_{ctrl,x}$$
$$C_m  =  C_{m α} α + C_{mp} p+ C_{mq} q+ C_{mr} r+ \sum_{x=1}^{CS} C_{m,ctrl,x} \delta_{ctrl,x}$$
$$C_n  =  C_{n β} β + C_{np} p+ C_{nq} q+ C_{nr} r+ \sum_{x=1}^{CS} C_{n,ctrl,x} \delta_{ctrl,x}$$
Here, the expression “CS” refers to the number of control surfaces on the aircraft. The summation adds up the effect of all the control surfaces.
Post stall, most of these equations have been left unchanged. The ones that do change compute the coefficients of lift and drag, which change to the following:
$$C_L = 2{sin} ^ {2} ( α )cos( α )+ C_{Lp} p+ C_{Lq} q+ C_{Lr} r+ \sum_{x=1}^{CS} C_{L,ctrl,x} \delta_{ctrl,x}$$

> $$C_D = C_{D,FP} (0.5-0.5cos(2 α ))+ C_{Dp} p+ C_{Dq} q+ C_{Dr} r+ \sum_{x=1}^{CS} C_{D,ctrl,x} \delta_{ctrl,x}$$

Reviewer: I'm not fully on board with leaving the stability and control derivatives as is in post stall regime. The stability derivatives can easily change sign, and control derivatives can lose their effectiveness. Without doing an actual post stall modeling of these - maybe the path of least resistance would be to just zero them out as post stall effects get ramped in? even if not totally accurate. Otherwise .. can be that e.g. control effectiveness still somewhat exists in stalled conditions.. but this is super airframe dependent. Maybe that an additional parameter for "stalled effectiveness" could be multiplied against the control surfaces if someone has this information for their airframe, or wants to qualitatively model this .. and then the default value is just zero for folks who dont.

other opinions / ideas? @bresch @Jaeyoung-Lim @huiyulhy @karthik-s-feather

Reviewer: @karthik-s-feather for this PR, probably ok to push it through without addressing this for now, but would still like to hear your thoughts on how you may address it in future PRs.

  $C_{D,FP} = \frac{2}{1+ e ^ {K1+K2 AR}}$
K1 and K2 are empirical coefficients. Currently, K1 is equal to -0.224, and K2 is equal to -0.115. <br> A sigmoid function is used to blend the pre-stall and post-stall models together, with a blending parameter M used to determine how sharply the plane stalls. Currently, the default value of M is 15, which switches from pre-stall to post-stall over the course of about 1.8 degrees of angle of attack.
$$\sigma = \frac{1+ e ^ {-M*( \alpha - \alpha_{stall} )} +e ^ {M*( \alpha - \alpha_{stall} )}}{(1+ e ^ {-M*( \alpha - \alpha_{stall} )} ) ( {1+e} ^ {M( α - α_{stall} )} )}$$
As such, the actual equations for the coefficient of lift and drag are:
$$C_L  =(1- \sigma )( C_{L0} + C_{L \alpha} \alpha )+ 2 \sigma {sin} ^ 2 ( α )cos( α )+ C_{Lp} p+ C_{Lq} q+ C_{Lr} r+ \sum_{x=1}^{CS} C_{L,ctrl,x} \delta_{ctrl,x}$$
$$C_D = (1-\sigma)(C_{D0}+\frac{{{C_L}} ^ {2}}{π ARe})+ \sigma C_{D,FP}(0.5-0.5cos(2 α ))+ C_{Dp} p+ C_{Dq} q+ C_{Dr} r+ \sum_{x=1}^{CS} C_{D,ctrl,x} \delta_{ctrl,x}$$
Given these coefficients, multiplying by the reference area of the aircraft (Sref) and the current dynamic pressure (q) produces the forces.

> $$F =  C_F S_{ref} q$$

Reviewer: I would probably use $\bar{q}$ for dynamic pressure to distinguish from $\hat{q}$ non-dimensionalized pitch rate and $q$ pitch rate

Author: Good point. Somehow there are three different variables in this file which typically are represented with "q". Check to make sure my new readme has all of those straightened out.

> $$q=\frac{1}{2} \rho v ^ {2}$$

Reviewer: define $\rho$ -- (air density)

The moments, however, need an additional term: either the span (b) or the mean aerodynamic chord (c) of the aircraft.
$$ℓ =  C_ℓ S_{ref} qb$$
$$m=  C_m S_{ref} qc$$
$$n =  C_n S_{ref} qb$$
<h1>Post-Stall Models</h1>

> As of August 2022, the plugin uses two post-stall models to compute the coefficient of lift and drag on the aircraft. The model for lift coefficient is a Newtonian model, based on the Third Law of Motion. 

Reviewer: ```suggestion
The plugin uses two post-stall models to compute the coefficient of lift and drag on the aircraft. The model for lift coefficient is a Newtonian model, based on the Third Law of Motion. 
```

> The post-stall drag model was constructed using two papers: one from Stringer et al which provides drag coefficients at angles of attack from zero to 360 degrees and the other from Ostowari and Naik which provides flat plate coefficients of drag at a variety of aspect ratios. The Stringer paper was used to relate drag to angle of attack, while Ostowari’s report was used to create a model of flat plate drag at a variety of aspect ratios. To create the latter model, I fit a sigmoid function to some of the numbers in the second paper, with the form below.

Reviewer: ```suggestion
The post-stall drag model was constructed using two papers: one from Stringer et al which provides drag coefficients at angles of attack from zero to 360 degrees and the other from Ostowari and Naik which provides flat plate coefficients of drag at a variety of aspect ratios. The Stringer paper was used to relate drag to angle of attack, while Ostowari’s report was used to create a model of flat plate drag at a variety of aspect ratios. To create the latter model, a sigmoid function is fit to some of the numbers in the second paper, with the form below.
```

> The post-stall drag model was constructed using two papers: one from Stringer et al which provides drag coefficients at angles of attack from zero to 360 degrees and the other from Ostowari and Naik which provides flat plate coefficients of drag at a variety of aspect ratios. The Stringer paper was used to relate drag to angle of attack, while Ostowari’s report was used to create a model of flat plate drag at a variety of aspect ratios. To create the latter model, I fit a sigmoid function to some of the numbers in the second paper, with the form below.

Reviewer: Can you please add links to the papers / books referenced here?

$$C_{D,fp} = \frac{2}{1+ e ^ {K1+K2 AR}}$$
The coefficients K1 and K2 are in the code, but might need some tuning.
A sigmoid function was used because CD initially increases quickly as aspect ratio rises, but that increase would slow down as AR goes to infinity. Intuitively, it seemed like a good model for the drag vs. aspect ratio relation, and the fit seemed pretty good as well.

> Looking at the data from Stringer, I could see that the CD was approximately equal to 1-cos(2α). This has a maximum value of 2, so I divided this expression by 2 to make sure multiplying it by flat-plate drag would not produce unreasonable results. 

Reviewer: ```suggestion
Looking at the data from Stringer, one can see that the CD was approximately equal to 1-cos(2α). This has a maximum value of 2, so here we divide this expression by 2 to make sure multiplying it by flat-plate drag would not produce unreasonable results. 
```
